server/api.server.js

```diff
@@ -199,3 +205,121 @@ async function waitForWebpack() {
     }
   }
 }
+
+const ignoreTypes = new Set();
+async function ssrMiddleware(req, res, next) {
+  // SSR only GET requests that define a ssr query param
+  if (req.method !== 'GET' || req.query.ssr !== '') {
+    return next();
+  }
+
+  await waitForWebpack();
+  const manifest = readFileSync(
+    path.resolve(__dirname, '../build/react-client-manifest.json'),
+    'utf8'
+  );
+  const moduleMap = JSON.parse(manifest);
+
+  const html = readFileSync(
+    path.resolve(__dirname, '../build/index.html'),
+    'utf8'
+  );
+  const [head, tail] = html.split('<div id="root">');
+  res.write(head + '<div id="root">');
+
+  // Instead of passing `res` (the response stream) to
+  // React's pipeToNodeWritable we pass a tranform `renderTransform` which is where we
+  // turn chunks into HTML.
+  // `renderTransform` is piped to `res`
+  //
+  // reactStream.pipe(renderTransform).pipe(res)
+  const renderTransform = new stream.Transform({
+    transform: (buf, enc, next) => {
+      const chunk = buf.toString();
+      const i = chunk.indexOf(':');
+      // Get type of message and id eg.
+      // J1 - where J stands for JSX
+      const [_, type, id] = chunk.slice(0, i).match(/(\w+)(\d+)/);
+
+      // This is a client module reference
+      if (type === 'M') {
+        // TODO: generate <link rel="preload" href="..." /> tag
+        return next(null, '');
+      }
+
+      // Other types include S (suspense) which we can ignore I think.
+      // ignoreTypes tracks references to other JSX elements that are in other chunks
+      // i.e. React suspended and later sent the resolved subtree in another chunk
+      // In SSR we will render a placeholder.
+      if (type !== 'J' || ignoreTypes.has(id)) {
+        return next(null, '');
+      }
+
+      // Render to string (HTML)
+      const json = JSON.parse(chunk.slice(i + 1));
+      const app = toReactTree(json);
+      next(null, ReactDOMServer.renderToStaticMarkup(app));
```

Comment:
probably this could also inline the "chunk" so that the client can grab it and stream it to React for hydration